### PR TITLE
8305972: Update XML Security for Java to 3.0.2

### PR DESCRIPTION
**Pull request by GoeLin**
I backport this for parity with 17.0.11-oracle.

The backport was almost clean, except for two trivial resolves due to differences in whitespace in the context.

The change comes with a CSR, which is already approved for 17.
But the CSR requires changes wrt. to the original change. 
In 17, no EDDSA support is added.

The PR comes with two commits:
1. the almost clean backport. I already skipped two comments added in head but not needed in 17 (SignatureMethod, DigestMethod).
2. removing the eddsa support.

Tests pass, SAP nightly testing passed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [ ] [JDK-8305972](https://bugs.openjdk.org/browse/JDK-8305972) needs maintainer approval
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8305972](https://bugs.openjdk.org/browse/JDK-8305972): Update XML Security for Java to 3.0.2 (**Enhancement** - P3)

### Reviewers
 * [Matthias Baesken](https://openjdk.org/census#mbaesken) (@MBaesken - **Reviewer**) ⚠️ Review applies to [5b271378](https://git.openjdk.org/jdk17u-dev/pull/2006/files/5b27137831084b58b929eb459e76874902864f35)
 * [Martin Balao](https://openjdk.org/census#mbalao) (@martinuy - **Reviewer**) ⚠️ Review applies to [3f4616cd](https://git.openjdk.org/jdk17u-dev/pull/2006/files/3f4616cd73744d31c8056fbb5c466282522ebde9)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk17u-dev.git pull/2006/head:pull/2006` \
`$ git checkout pull/2006`

Update a local copy of the PR: \
`$ git checkout pull/2006` \
`$ git pull https://git.openjdk.org/jdk17u-dev.git pull/2006/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 2006`

View PR using the GUI difftool: \
`$ git pr show -t 2006`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk17u-dev/pull/2006.diff">https://git.openjdk.org/jdk17u-dev/pull/2006.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk17u-dev/pull/2006#issuecomment-1833349699)

### CODE REVIEW

**Review by MBaesken — COMMENTED** [review]
[...ava.xml.crypto/share/classes/com/sun/org/apache/xml/internal/security/resource/config.xml](https://github.com/openjdk/jdk17u-dev/pull/2006/files#diff-5eb925fbd6a3b144425f263e233ded60235050e04cc36e47cac7c9a9bdd97008)

that looks strange, is that really intended ?
Unless required by applicable law or agreed to in writing,ons.SignatureBaseRSA$SignatureRSASHA3_512MGF1" />

**Review by MBaesken — APPROVED** [review]
The deleted line in [src/java.xml.crypto/share/classes/org/jcp/xml/dsig/internal/dom/DOMSignatureMethod.java](https://github.com/openjdk/jdk17u-dev/pull/2006/files#diff-cdae7f9a1a033faaeeb5482bcbe882149c3881acf409558f5c7297a5ef3f205b)
might be restored but otherwise looks okay to me.

File: src/java.xml.crypto/share/classes/org/jcp/xml/dsig/internal/dom/AbstractDOMSignatureMethod.java
[the remark this answers is not in the document]
Comment (GoeLin) [reply src/java.xml.crypto/share/classes/org/jcp/xml/dsig/internal/dom/AbstractDOMSignatureMethod.java:1]: Fixed, thanks

File: src/java.xml.crypto/share/classes/org/jcp/xml/dsig/internal/dom/DOMSignatureMethod.java [hunk 1]
```diff
@@ -1020,5 +1020,4 @@ Type getAlgorithmType() {
             return Type.ECDSA;
         }
     }
-
```
[the remark this answers is not in the document]
Comment (GoeLin) [reply src/java.xml.crypto/share/classes/org/jcp/xml/dsig/internal/dom/DOMSignatureMethod.java:1023]: Fixed, thanks

**Review by martinuy — APPROVED** [review]
Hi Goetz,

Thanks for proposing this backport.

One minor comment:

 * test/jdk/javax/xml/crypto/dsig/GenerationTests.java
  * The import of X509Certificate and the definition of x5ks are dead code.

Otherwise, looks good to me. I'll give my approval anyways, as this is a minor comment.

On a final note, I agree with your changes of removing EdDSA code and aligning to the approved CSR. However, I have to say that doing this for 17u does not only prevent users from the enhancement but also increases the maintenance cost as there will be more chances of updates not applying cleanly for the years to come. I would have treated 17u differently than previous releases.

Martin.-

**Comment by GoeLin** [comment]
Hi @martinuy,
do you think it would be better to keep all the code and only remove the two String and the change to the comment of DigestMethod?
In the 21u change for the update to [3.0.3](https://github.com/openjdk/jdk21u-dev/pull/94) I have done it that way. Probably that's also the better solution here.

**Comment by martinuy** [comment]
Hi @GoeLin ,

I assume that in 21u you kept all the (implementation) code except for the public members. If so, I understand the motivations but personally prefer what you proposed for 17u in this PR. It makes the code more clear in terms of what is supported. For example, it would be misleading for someone who looks for "EdDSA" references in the code, finds many —even beyond defines— and assumes that it is supported. This is, of course, at the expense of higher chances of non-clean updates. We have taken this approach in other libraries before such as when we removed the implementation of DTLS in the 8u backport of the TLS engine.

Martin.-

**Comment by GoeLin** [comment]
Hi @martinuy,
are you fine with me driving [#2116](https://github.com/openjdk/jdk17u-dev/pull/2116) and similarly [#2455](https://github.com/openjdk/jdk11u-dev/pull/2455)  for 11?

**Comment by martinuy** [comment]
Hi @GoeLin ,

I personally prefer the approach taken in PR #2006 (here), for the reasons in my [previous comment](https://github.com/openjdk/jdk17u-dev/pull/2006#issuecomment-1881457955). With that said, I see your motivation. I would be interested in hearing @jerboaa view on this issue.

**Comment by jerboaa** [comment]
From the 17, 11 and 8u CSR:

```
The port is largely the same in terms of implementation as that done for JDK 21.

The javax.xml.crypto.dsig.SignatureMethod and javax.xml.crypto.dsig.DigestMethod interfaces will not be updated. Instead, end users would define the newly added EdDSA Signature methods locally in application code. Unlike JDK 17 and later, JDK 11 and 8 doesn't have EdDSA support by default. A 3rd party security provider which supports ed25519 and ed448 would be required.
```

This suggests that EdDSA support was kept as this bug is about updating a bundled in-tree library from a third party (Apache Santuario). So changes to that third party code should be kept to a minimum, IMO. For JDK 11 and JDK 8 even a third party provider would be needed in order to be able to use EdDSA there.

TLDR; Since this is a third-party library code upgrade, I'd suggest to go with https://github.com/openjdk/jdk17u-dev/pull/2116

HTH.

**Comment by GoeLin** [comment]
Thanks for your help! Closing this PR.
